This exercise will require you to pull some data from https://data.nasdaq.com/ (formerly Quandl API).

As a first step, you will need to register a free account on the https://data.nasdaq.com/ website.

After you register, you will be provided with a unique API key, that you should store:

*Note*: Use a `.env` file and put your key in there and `python-dotenv` to access it in this notebook. 

The code below uses a key that was used when generating this project but has since been deleted. Never submit your keys to source control. There is a `.env-example` file in this repository to illusrtate what you need. Copy that to a file called `.env` and use your own api key in that `.env` file. Make sure you also have a `.gitignore` file with a line for `.env` added to it. 

The standard Python gitignore is [here](https://github.com/github/gitignore/blob/master/Python.gitignore) you can just copy that. 

Nasdaq Data has a large number of data sources, but, unfortunately, most of them require a Premium subscription. Still, there are also a good number of free datasets.

For this mini project, we will focus on equities data from the Frankfurt Stock Exhange (FSE), which is available for free. We'll try and analyze the stock prices of a company called Carl Zeiss Meditec, which manufactures tools for eye examinations, as well as medical lasers for laser eye surgery: https://www.zeiss.com/meditec/int/home.html. The company is listed under the stock ticker AFX_X.

You can find the detailed Nasdaq Data API instructions here: https://docs.data.nasdaq.com/docs/in-depth-usage

While there is a dedicated Python package for connecting to the Nasdaq API, we would prefer that you use the *requests* package, which can be easily downloaded using *pip* or *conda*. You can find the documentation for the package here: http://docs.python-requests.org/en/master/ 

Finally, apart from the *requests* package, you are encouraged to not use any third party Python packages, such as *pandas*, and instead focus on what's available in the Python Standard Library (the *collections* module might come in handy: https://pymotw.com/3/collections/).
Also, since you won't have access to DataFrames, you are encouraged to us Python's native data structures - preferably dictionaries, though some questions can also be answered using lists.
You can read more on these data structures here: https://docs.python.org/3/tutorial/datastructures.html

Keep in mind that the JSON responses you will be getting from the API map almost one-to-one to Python's dictionaries. Unfortunately, they can be very nested, so make sure you read up on indexing dictionaries in the documentation provided above.

Note: API's can change a bit with each version, for this exercise it is reccomended to use the nasdaq api at `https://data.nasdaq.com/api/v3/`. This is the same api as what used to be quandl so `https://www.quandl.com/api/v3/` should work too.

Hint: We are looking for the `AFX_X` data on the `datasets/FSE/` dataset.

In [1]:
# get api key from your .env file
import os
from dotenv import load_dotenv

load_dotenv()
#confused, how does it know which file? 
#what if you have multiple? 
#what was the point of gitignore?
API_KEY = os.getenv('NASDAQ_API_KEY') 
# guessing this is because they are a key val pair, so asking for the key gives the val

# print(API_KEY)

In [2]:
# import relevant modules
import collections
import requests

These are your tasks for this mini project:

1. Collect data from the Franfurt Stock Exchange, for the ticker AFX_X, for the whole year 2017 (keep in mind that the date format is YYYY-MM-DD).
2. Convert the returned JSON object into a Python dictionary.
3. Calculate what the highest and lowest opening prices were for the stock in this period.
4. What was the largest change in any one day (based on High and Low price)?
5. What was the largest change between any two days (based on Closing Price)?
6. What was the average daily trading volume during this year?
7. (Optional) What was the median trading volume during this year. (Note: you may need to implement your own function for calculating the median.)

In [3]:
url = 'https://data.nasdaq.com/api/v3/datasets/FSE/AFX_X'
dataset = 'datasets/FSE/'
data = 'AFX_X'

In [4]:
# Now, call the Nasdaq API and pull out a small sample of the data (only one day) to get a glimpse 
# into the JSON structure that will be returned

# Task 1
param =  dict(key=API_KEY)
res = requests.get(url, params=param)

# Task 2
json_data = res.json()

In [5]:
# Inspect the JSON structure of the object you created, and take note of how nested it is,
# as well as the overall structure

In [6]:
type(json_data)

dict

In [7]:
json_data.keys()

dict_keys(['dataset'])

#### NOTES
for key, value in json_data.items():
    print(key + ':', value)

In [28]:
# json_data['dataset']

In [9]:
json_data['dataset']['column_names']

['Date',
 'Open',
 'High',
 'Low',
 'Close',
 'Change',
 'Traded Volume',
 'Turnover',
 'Last Price of the Day',
 'Daily Traded Units',
 'Daily Turnover']

In [10]:
json_data['dataset']['column_names'][1]

'Open'

In [11]:
x = json_data['dataset']['data'][0]
type(x[0])

str

In [12]:
# YYYY-MM-DD
start = '2017-01-01'
end = '2017-12-31'

open_list = []
high_list = []
low_list = []
close_list = []
volume_list = []
for value in json_data['dataset']['data']:
    if value[0] >= start and value[0] <= end:
        open_list.append(value[1])
        high_list.append(value[2])
        low_list.append(value[3])
        close_list.append(value[4])
        volume_list.append(value[6])

In [13]:
len(high_list)

255

In [14]:
len(high_list) == len(low_list)

True

### Task 3
Calculate what the highest and lowest opening prices were for the stock in this period.

In [15]:
# Task 3
clean_open = []
for val in open_list:
    if val != None:
        clean_open.append(val)
        
len(clean_open)

252

In [16]:
(max(clean_open),min(clean_open))

(53.11, 34.0)

The highest opening price for stock in this period is 53.0 and the lowest is 34.0.

### Task 4
What was the largest change in any one day (based on High and Low price)?

In [17]:
# CHECK FOR NONE VALUES
clean_high = []
for val in high_list:
    if val != None:
        clean_high.append(val)
        
len(clean_high) == len(high_list)

True

In [18]:
# Task 4
high_low = zip(high_list, low_list)
high_low_list = []
for el1, el2 in high_low:
    high_low_list.append(el1 - el2)

In [19]:
len(high_low_list)

255

In [20]:
max(high_low_list)

2.8100000000000023

The largest change in any one day based on High and Low prices is 2.81.

### Task 5
What was the largest change between any two days (based on Closing Price)?

In [21]:
# CHECK FOR NONE VALUES
# close_list
clean_close = []
for val in close_list:
    if val != None:
        clean_close.append(val)
        
len(clean_close) == len(close_list)

True

In [22]:
x_close = []
for ind in range(len(close_list)-1):
#while ind < (len(close_list)-1):
    x_close.append(close_list[ind+1]-close_list[ind])

In [23]:
# check
(close_list[45]-close_list[44]) == x_close[44]

True

In [24]:
max(x_close)

2.559999999999995

The largest change between any two days based on Closing Price is 2.56.

### Task 6
What was the average daily trading volume during this year?

In [25]:
sum(volume_list)/len(volume_list)

89124.33725490196

The average daily trading volume in 2017 was 89,124.34.

### Task 7
(Optional) What was the median trading volume during this year. (Note: you may need to implement your own function for calculating the median.)

In [26]:
len(volume_list) % 2 

1

In [27]:
volume_list.sort()
quotient, remainder = divmod(len(volume_list), 2)
if remainder:
    med_volume = volume_list[quotient]
else:
    med_volume = sum(volume_list[quotient - 1:quotient + 1]) / 2
med_volume

76286.0

The median trading volume during 2017 was 76286.